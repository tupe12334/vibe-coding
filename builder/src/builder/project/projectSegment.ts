import { readFile } from "fs/promises";
import { join } from "path";

export const projectSegment = async (
  templatesPath: string,
  projectType: string
) => {
  const projectJsonFile = (
    await readFile(join(templatesPath, "project", `${projectType}.json`), {
      encoding: "utf-8",
    })
  ).toString();
  const projectItems = JSON.parse(projectJsonFile);

  const projectSegment = [
    {
      type: "heading",
      depth: 2,
      children: [
        { type: "text", value: "Project type" },
        { type: "text", value: ` (${projectType})` },
      ],
    },
    ...projectItems.map((item: string) => ({
      type: "listItem",
      children: [{ type: "text", value: item }],
    })),
  ];

  return projectSegment;
};
