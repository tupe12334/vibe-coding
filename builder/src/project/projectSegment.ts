import { readFile } from "fs/promises";
import { join } from "path";
import type { RootContent } from "mdast";

export const projectSegment = async (
  templatesPath: string,
  projectType: string
): Promise<RootContent[]> => {
  const projectJsonFile = (
    await readFile(join(templatesPath, "project", `${projectType}.json`), {
      encoding: "utf-8",
    })
  ).toString();
  const projectItems = JSON.parse(projectJsonFile);

  const projectSegment: RootContent[] = [
    {
      type: "heading",
      depth: 2,
      children: [
        { type: "text", value: "Project type" },
        { type: "text", value: ` (${projectType})` },
      ],
    },
    ...projectItems.map((item: string) => ({
      type: "listItem",
      children: [{ type: "text", value: item }],
    })),
  ];

  return projectSegment;
};
