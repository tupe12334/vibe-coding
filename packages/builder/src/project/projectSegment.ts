import type { RootContent } from "mdast";
import { backendRules } from "./backend";
import { frontendRules } from "./frontend";
import { libRules } from "./lib";
import { uiLibRules } from "./ui-lib";
import { e2eRules } from "./e2e";
import { ProjectTypes } from "./options";

const projectRulesMap: Record<string, readonly string[]> = {
  backend: backendRules,
  frontend: frontendRules,
  lib: libRules,
  "ui-lib": uiLibRules,
  e2e: e2eRules,
};

export const projectSegment = async (
  projectType: string
): Promise<RootContent[]> => {
  let projectItems = projectRulesMap[projectType] ?? [];

  const projectConfig = ProjectTypes.find((p) => p.name === projectType);
  if (projectConfig?.subset) {
    for (const subset of projectConfig.subset) {
      const subsetItems = projectRulesMap[subset] ?? [];
      projectItems = projectItems.concat(subsetItems);
    }
  }

  const projectSegment: RootContent[] = [
    {
      type: "heading",
      depth: 2,
      children: [
        { type: "text", value: "Project type" },
        { type: "text", value: ` (${projectType})` },
      ],
    },
    {
      type: "list",
      ordered: false,
      children: projectItems.map((item: string) => ({
        type: "listItem",
        children: [
          {
            type: "paragraph",
            children: [{ type: "text", value: item }],
          },
        ],
      })),
    },
  ];

  return projectSegment;
};
