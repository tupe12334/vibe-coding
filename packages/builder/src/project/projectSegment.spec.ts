import { describe, expect, test } from "vitest";
import { projectSegment } from "./projectSegment";

describe("projectSegment", () => {
  test("ui-lib includes lib rules", async () => {
    const segment = await projectSegment("ui-lib");
    expect(segment).toMatchSnapshot();
  });
});
