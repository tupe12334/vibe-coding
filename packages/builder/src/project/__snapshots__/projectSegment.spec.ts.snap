// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`projectSegment > ui-lib includes lib rules 1`] = `
[
  {
    "children": [
      {
        "type": "text",
        "value": "Project type",
      },
      {
        "type": "text",
        "value": " (ui-lib)",
      },
    ],
    "depth": 2,
    "type": "heading",
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Prefer working with flex",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Prioritize the use of components that exist in the library already.",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Folder structure:

- \`src/\`: Contains the source code for the components.
- \`src/<component>/\`: Contains the individual components.
  - Each component has its own folder with the following structure:
    - \`index.ts\`: Exports all the relevant data.
    - \`<ComponentName>.tsx\`: Contains the main component code.
    - \`styles.ts\`: Contains styles for the component.
    - \`types.ts\`: Contains TypeScript types for the component.
    - \`<ComponentName>.test.tsx\`: Contains tests for the component.
    - \`<ComponentName>.stories.tsx\`: Contains Storybook stories for the component.
- \`src/index.ts\`: Exports all components for easy import.",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Work with version control for publish the package if its publishable",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Create a clear and comprehensive README.md with installation instructions, usage examples, and API documentation",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Use semantic versioning (semver) for package versions",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Include proper TypeScript declaration files (.d.ts) for better developer experience",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Set up automated testing with good test coverage before publishing",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Configure proper entry points in package.json (main, module, types fields)",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Consider tree-shaking compatibility by using ES modules",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "children": [
          {
            "children": [
              {
                "type": "text",
                "value": "Add proper keywords and description in package.json for discoverability",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
    ],
    "ordered": false,
    "type": "list",
  },
]
`;
